File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/study/StudyViewModel.kt
package com.jworks.kanjiquest.android.ui.study

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.android.data.PreviewTrialManager
import com.jworks.kanjiquest.core.domain.UserSessionProvider
import com.jworks.kanjiquest.core.domain.model.FlashcardDeckGroup
import com.jworks.kanjiquest.core.domain.model.GameMode
import com.jworks.kanjiquest.core.domain.model.UserLevel
import com.jworks.kanjiquest.core.domain.repository.FlashcardRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class StudyUiState(
    val selectedTab: ContentTab = ContentTab.KANJI,
    val selectedMode: GameMode? = null,
    val selectedSource: StudySource = StudySource.All,
    val kanaFilter: KanaFilter = KanaFilter.HIRAGANA_ONLY,
    val isPremium: Boolean = false,
    val isAdmin: Boolean = false,
    val decks: List<FlashcardDeckGroup> = emptyList(),
    val previewTrialsRemaining: Map<GameMode, Int> = emptyMap()
)

@HiltViewModel
class StudyViewModel @Inject constructor(
    private val flashcardRepository: FlashcardRepository,
    private val userSessionProvider: UserSessionProvider,
    private val previewTrialManager: PreviewTrialManager
) : ViewModel() {

    private val _uiState = MutableStateFlow(StudyUiState())
    val uiState: StateFlow<StudyUiState> = _uiState.asStateFlow()

    init {
        loadData()
    }

    private fun loadData() {
        viewModelScope.launch {
            val effectiveLevel = userSessionProvider.getEffectiveLevel()
            val isPremium = effectiveLevel == UserLevel.PREMIUM || effectiveLevel == UserLevel.ADMIN
            val decks = try { flashcardRepository.getAllDeckGroups() } catch (_: Exception) { emptyList() }

            val trials = mapOf(
                GameMode.WRITING to previewTrialManager.getRemainingTrials(GameMode.WRITING),
                GameMode.VOCABULARY to previewTrialManager.getRemainingTrials(GameMode.VOCABULARY),
                GameMode.CAMERA_CHALLENGE to previewTrialManager.getRemainingTrials(GameMode.CAMERA_CHALLENGE)
            )

            _uiState.value = _uiState.value.copy(
                isPremium = isPremium,
                isAdmin = userSessionProvider.isAdmin(),
                decks = decks,
                previewTrialsRemaining = trials,
                selectedMode = getDefaultModeForTab(_uiState.value.selectedTab)
            )
        }
    }

    fun selectTab(tab: ContentTab) {
        _uiState.value = _uiState.value.copy(
            selectedTab = tab,
            selectedMode = getDefaultModeForTab(tab),
            selectedSource = StudySource.All
        )
    }

    fun selectMode(mode: GameMode) {
        _uiState.value = _uiState.value.copy(selectedMode = mode)
    }

    fun selectSource(source: StudySource) {
        _uiState.value = _uiState.value.copy(selectedSource = source)
    }

    fun selectKanaFilter(filter: KanaFilter) {
        _uiState.value = _uiState.value.copy(kanaFilter = filter)
    }

    fun usePreviewTrial(mode: GameMode): Boolean {
        val success = previewTrialManager.usePreviewTrial(mode)
        if (success) {
            _uiState.value = _uiState.value.copy(
                previewTrialsRemaining = _uiState.value.previewTrialsRemaining.toMutableMap().apply {
                    this[mode] = previewTrialManager.getRemainingTrials(mode)
                }
            )
        }
        return success
    }

    private fun getDefaultModeForTab(tab: ContentTab): GameMode {
        return when (tab) {
            ContentTab.KANA -> GameMode.KANA_RECOGNITION
            ContentTab.RADICALS -> GameMode.RADICAL_RECOGNITION
            ContentTab.KANJI -> GameMode.RECOGNITION
        }
    }

    fun getAvailableModes(tab: ContentTab): List<GameMode> {
        return when (tab) {
            ContentTab.KANA -> listOf(GameMode.KANA_RECOGNITION, GameMode.KANA_WRITING)
            ContentTab.RADICALS -> listOf(GameMode.RADICAL_RECOGNITION)
            ContentTab.KANJI -> listOf(GameMode.RECOGNITION, GameMode.WRITING, GameMode.VOCABULARY, GameMode.CAMERA_CHALLENGE)
        }
    }

    fun isModeAccessible(mode: GameMode): Boolean {
        val state = _uiState.value
        if (state.isPremium || state.isAdmin) return true
        return when (mode) {
            GameMode.RECOGNITION, GameMode.KANA_RECOGNITION, GameMode.KANA_WRITING,
            GameMode.RADICAL_RECOGNITION -> true
            GameMode.WRITING -> (state.previewTrialsRemaining[GameMode.WRITING] ?: 0) > 0
            GameMode.VOCABULARY -> (state.previewTrialsRemaining[GameMode.VOCABULARY] ?: 0) > 0
            GameMode.CAMERA_CHALLENGE -> (state.previewTrialsRemaining[GameMode.CAMERA_CHALLENGE] ?: 0) > 0
            else -> false
        }
    }
}

File: ios-app/KanjiQuest/Screens/Study/StudyViewModel.swift
import Foundation
import SharedCore

/// UI state matching Android's StudyUiState.
struct StudyUiState {
    var selectedTab: ContentTab = .kanji
    var selectedMode: GameModeEnum? = nil
    var selectedSource: StudySource = .all
    var kanaFilter: KanaFilter = .hiraganaOnly
    var isPremium: Bool = false
    var isAdmin: Bool = false
    var previewTrialsRemaining: [GameModeEnum: Int] = [:]
    var decks: [(id: Int64, name: String)] = []
}

/// ViewModel matching Android's StudyViewModel.
@MainActor
class StudyViewModel: ObservableObject {
    @Published var uiState = StudyUiState()

    private let container: AppContainer

    init(container: AppContainer) {
        self.container = container
        Task { await loadData() }
    }

    func loadData() async {
        let effectiveLevel = container.userSessionProvider.getEffectiveLevel()
        let isPremium = effectiveLevel == .premium || effectiveLevel == .admin
        let isAdmin = container.userSessionProvider.isAdmin()

        // Load flashcard decks
        let decksList: [(id: Int64, name: String)]
        if let decks = try? await container.flashcardRepository.getAllDeckGroups() {
            decksList = decks.map { (id: $0.id, name: $0.name) }
        } else {
            decksList = []
        }

        // Load preview trial counts
        var trials: [GameModeEnum: Int] = [:]
        for mode in GameModeEnum.allCases {
            if let key = mode.previewTrialKey {
                trials[mode] = container.previewTrialManager.getRemainingTrials(mode: key)
            }
        }

        uiState.isPremium = isPremium
        uiState.isAdmin = isAdmin
        uiState.decks = decksList
        uiState.previewTrialsRemaining = trials
    }

    func selectTab(_ tab: ContentTab) {
        uiState.selectedTab = tab
        uiState.selectedMode = nil
        uiState.selectedSource = .all
    }

    func selectMode(_ mode: GameModeEnum) {
        uiState.selectedMode = mode
    }

    func selectSource(_ source: StudySource) {
        uiState.selectedSource = source
    }

    func selectKanaFilter(_ filter: KanaFilter) {
        uiState.kanaFilter = filter
    }

    func isModeAccessible(_ mode: GameModeEnum) -> Bool {
        if uiState.isPremium || uiState.isAdmin { return true }
        if !mode.isPremiumGated { return true }
        // Check preview trials
        if let key = mode.previewTrialKey {
            return container.previewTrialManager.getRemainingTrials(mode: key) > 0
        }
        return false
    }

    func usePreviewTrial(_ mode: GameModeEnum) {
        if let key = mode.previewTrialKey {
            _ = container.previewTrialManager.usePreviewTrial(mode: key)
            uiState.previewTrialsRemaining[mode] = container.previewTrialManager.getRemainingTrials(mode: key)
        }
    }
}
